workloads/files/workload-network-script-cm.yml

```diff
@@ -0,0 +1,961 @@
+            containers:
+            - name: "uperf-\${NS_VALUE}"
+              image: ${NETWORK_TEST_UPERF_IMAGE}
+              imagePullPolicy: Always
+              env:
+              - name: UPERF_SSH_PORT
+                value: \"${NETWORK_TEST_UPERF_SSHD_PORT}\"
+              securityContext:
+                privileged: ${NETWORK_TEST_HOSTNETWORK}
+              volumeMounts:
+              - name: pbench-ssh
+                mountPath: /.ssh/authorized_keys
+                subPath: authorized_keys
+              - name: pbench-ssh
+                mountPath: /root/.ssh/authorized_keys
+                subPath: authorized_keys
+              - name: pbench-results
+                mountPath: /var/lib/pbench-agent
+            dnsPolicy: ClusterFirst
+            nodeSelector:
+              uperf: "\${NS_VALUE}"
+            restartPolicy: Always
+            volumes:
+            - name: pbench-ssh
+              secret:
+                secretName: pbench-ssh
+                defaultMode: 0600
+            - name: pbench-results
+              emptyDir: {}
+    parameters:
+    - name: NS_VALUE
+      description: Which NodeSelector value to use (client, server)
+      value: primary
+    - name: IDENTIFIER
+      description: Number to append to the name of resources
+      value: '1'
+      required: true
+  pbench-ssh.yaml.template: |
+    ---
+    Kind: Template
+    apiVersion: v1
+    metadata:
+      name: sshSecretTemplate
+      creationTimestamp:
+      annotations:
+        description: This template will create a single secret.
+        tags: ''
+    objects:
+      - apiVersion: v1
+        kind: Secret
+        metadata:
+          name: pbench-ssh
+        type: Opaque
+        data:
+          authorized_keys: ${NETWORK_SSH_AUTHORIZED_KEYS}
+          id_rsa: ${NETWORK_SSH_PRIVATE_KEY}
+          id_rsa.pub: ${NETWORK_SSH_PUBLIC_KEY}
+    parameters:
+    - name: IDENTIFIER
+      description: Number to append to the name of resources
+      value: '1'
+      required: true
+  ssh_config.template: |
+    Host ${host}
+      User default
+      Port ${NETWORK_TEST_UPERF_SSHD_PORT}
+      StrictHostKeyChecking no
+      PasswordAuthentication no
+      UserKnownHostsFile /dev/null
+      IdentityFile ~/.ssh/id_rsa
+  svc-template.yaml.template: |
+    ---
+    kind: Template
+    apiVersion: v1
+    metadata:
+      name: uperf
+      annotations:
+        description: A template for creating resources for network testing
+        tags: uperf,networking
+    labels:
+      template: uperf
+    objects:
+    - apiVersion: v1
+      kind: Service
+      metadata:
+        labels:
+          name: "\${ROLE}"
+          svc_selector: "\${ROLE}-\${IDENTIFIER}"
+        annotations:
+          description: Exposes and load balances the application pods
+        name: uperf-\${ROLE}-\${IDENTIFIER}
+      spec:
+        ports:
+        - name: ssh
+          port: ${NETWORK_TEST_UPERF_SSHD_PORT}
+        - name: tcp-20000
```

Comment:
Cannot we use a range of some sort?  If not, can this be templated?  So many ports copied, there must be another way.

Reply:
I wish there was a range but I couldn't find any implementation of a port range but I did find this open kubernetes issue - https://github.com/kubernetes/kubernetes/issues/23864

Each pair of uperf clients needs one port per uperf "instance" thus the first pair can use 20010-20019, 2nd pair 20020-20029, ...

It actually gets worse, uperf needs a control port opened somewhere between 35000-45000(?) which is problematic for kubernetes services since each port needs to be spelled out.  I am still in the process of understanding how this control port was introduced into pbench-uperf and if we can add a mechanism to control which port it uses so we can precisely control it here.

In the meantime I'll see about moving the ports to be programmatically added.

Reply:
Wouldn't jinja2 for loop be of any help?